ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,559 @@
+                        ver: PhantomData,
+                        key: 0,
+                    })
+                }
+            }
+        }
+        Ok(iter)
+    }
+}
+
+impl<'a, V, BS, Ver> IntoIterator for &'a crate::AmtImpl<V, BS, Ver>
+where
+    V: Serialize + DeserializeOwned,
+    Ver: crate::root::version::Version,
+    BS: Blockstore,
+{
+    type IntoIter = Iter<'a, V, &'a BS, Ver>;
+    type Item = Result<(usize, &'a V), crate::Error>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct Iter<'a, V, BS, Ver> {
+    stack: Vec<IterStack<'a, V>>,
+    blockstore: BS,
+    bit_width: u32,
+    ver: PhantomData<Ver>,
+    key: usize,
+}
+
+pub struct IterStack<'a, V> {
+    pub(crate) node: Option<&'a Node<V>>,
+    pub(crate) idx: usize,
+}
+
+impl<'a, V, BS, Ver> Iterator for Iter<'a, V, BS, Ver>
+where
+    BS: Blockstore,
+    V: Serialize + DeserializeOwned,
+{
+    type Item = Result<(usize, &'a V), crate::Error>;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let stack = self.stack.last_mut()?;
+            match stack.node {
+                Some(Node::Leaf { vals }) => {
+                    while stack.idx < vals.len() {
+                        match vals[stack.idx] {
+                            Some(ref v) => {
+                                stack.idx += 1;
+                                self.key += 1;
+                                return Some(Ok((self.key - 1, v)));
+                            }
+                            None => {
+                                stack.idx += 1;
+                                self.key += 1;
+                            }
+                        }
+                    }
+                    self.stack.pop();
+                }
+                Some(Node::Link { links }) => {
+                    if stack.idx < links.len() {
+                        let link = &links[stack.idx];
+                        match link {
+                            Some(Link::Cid { cid, cache }) => {
+                                match cache.get_or_try_init(|| {
+                                    self.blockstore
+                                        .get_cbor::<CollapsedNode<V>>(cid)?
+                                        .ok_or_else(|| Error::CidNotFound(cid.to_string()))?
+                                        .expand(self.bit_width)
+                                        .map(Box::new)
+                                }) {
+                                    Ok(node) => {
+                                        stack.idx += 1;
+                                        self.stack.push(IterStack {
+                                            node: Some(node.as_ref()),
+                                            idx: 0,
+                                        });
+                                    }
+                                    Err(e) => return Some(Err(e)),
```

Comment:
Ideally we'd increment the index, even on error (e.g., so we can skip over missing sections).